ceph: bucket notifications refactoring work

yuvalif opened the pull request:
This PR is addressing followup work from: https://github.com/rook/rook/pull/8426

- [x] code reuse in the bucket notification controller and OBC label controller
- [x] use the regular S3 agent for notification provisioning (topic will still need the special SNS client)
- [x] move the OBC label predicate to `WatchPredicateForNonCRDObject`
- [x] go through all of the controllers' reconciliation error conditions and decide which should return error with nil `ReportReconcileResult` (immediate retry) and which should define a delayed retry by returning error with `ReportReconcileResult`
- [x] add unit tests for the notification controller
- [x] unit tests for the obc label controller
- [x] add unit tests for the topic controller

Following tasks will be broken off to their own PRs:
- add OBC list in notification status
- fetch the list of notifications from the RGW and do the diff with the labels instead of deleting all notifications on any label change
- More integration tests: https://github.com/rook/rook/pull/9169

leseb commented on pkg/operator/ceph/object/notification/controller.go:
```diff
@@ -237,26 +197,46 @@ func (r *ReconcileNotifications) updateCephBucketNotification(notification *ceph
 	return reconcile.Result{}, nil
 }
+
+func getCephObjectStoreName(ob bktv1alpha1.ObjectBucket) (objectStoreName types.NamespacedName, err error) {
```
```suggestion
func getCephObjectStoreName(ob bktv1alpha1.ObjectBucket) (types.NamespacedName, error) {
```

While we are at it, can we stop naming returns? We very rarely do it in our code base and prefer returning like


```go
return objectStoreName, nil
```

yuvalif replied:
sure, can change that.
as i'm not "native" in golang :-) is there a general guideline to avoid that?
or is this just a specific style n rook?

leseb replied:
There is no general guidelines to avoid this, and actually, effective go "recommends" it or at least says it can be useful for clarity: https://golang.org/doc/effective_go#named-results. However, I haven't seen many projects doing it and Rook certainly does not. I believe we have 1 exception that works with a `defer` statement and that's it.

yuvalif replied:
ok. will follow the "When in Rome.." guideline

yuvalif replied:
i cleaned that up in other places in the code. kepts it in one place, where it actually add clarity. I have a function:
```golang
func getUserCredentials(...) (accessKey string, secretKey string, err error)
```
where naming the return strings allows the caller to understand what are the return values without looking into the implementation of the function

leseb commented on pkg/operator/ceph/controller/predicate.go:
```diff
@@ -364,24 +364,6 @@ func WatchControllerPredicate() predicate.Funcs {
 					return true
 				}
-
-			case *bktv1alpha1.ObjectBucketClaim:
```
Why moving this code out to `WatchPredicateForNonCRDObject`?

yuvalif replied:
was requested by @BlaineEXE

leseb replied:
@BlaineEXE do you remember why?

leseb replied:
This is a CRD so its right place is in `WatchControllerPredicate()` IMO. Especially since it's called via `&source.Kind{Type: &bktv1alpha1.ObjectBucketClaim{}}` in the watcher.

yuvalif commented on pkg/operator/ceph/object/notification/controller.go:
```diff
@@ -160,52 +155,17 @@ func (r *ReconcileNotifications) reconcile(request reconcile.Request) (reconcile
 
 	// Populate clusterInfo during each reconcile
-	r.clusterInfo, _, _, err = mon.LoadClusterInfo(r.context, r.opManagerContext, cephCluster.Namespace)
+	clusterInfo, _, _, err := mon.LoadClusterInfo(r.context, r.opManagerContext, cephCluster.Namespace)
```
it is not needed in the `ReconcileNotifications` struct
removed from line 58

leseb commented on pkg/operator/ceph/object/topic/controller_test.go:
```diff
@@ -0,0 +1,245 @@
+/*
+Copyright 2020 The Rook Authors. All rights reserved.
```
```suggestion
Copyright 2021 The Rook Authors. All rights reserved.
```

leseb commented on pkg/operator/ceph/object/topic/provisioner.go:
```diff
@@ -142,17 +138,12 @@ func createSNSClient(sess *awssession.Session) (snsClient *sns.SNS) {
 }
 
-func (p *Provisioner) Create(topic *cephv1.CephBucketTopic) (*string, error) {
+var Create = func(p *Provisioner, topic *cephv1.CephBucketTopic) (*string, error) {
```
Do we really need that pointer to a string in the return?

Also why is `p` now a pointer? Typically, if it's a small struct and we don't need to mutate it, it shouldn't be a pointer.

yuvalif replied:
> Do we really need that pointer to a string in the return?

the ARN in the `CreateTopicOutput` (an AWS SDK struct) is a `*string`, so I'm just returning the value from there

thotz commented on pkg/operator/ceph/controller/predicate.go:
```diff
@@ -528,6 +516,32 @@ func WatchPredicateForNonCRDObject(owner runtime.Object, scheme *runtime.Scheme)
+						logger.Infof("labels has changed for %q. diff=%s", newObj.GetName(), diff)
+						return true
+					}
+					if oldObj.Spec.ObjectBucketName != newObj.Spec.ObjectBucketName {
```
AFAIR only Spec can be modified for OBC is the `AdditionalConfig` field others are immutable

yuvalif replied:
maybe it should not be done, but when the bucket is actually created this is changing

thotz replied:
Okay `Spec.ObjectBucketName` is filled for referring the `ob Name`. IMO below case checks for status and if it is `bounded` state then backend bucket will be created. There is no harm in having this check but the below checks gives the required validation

yuvalif replied:
when i ran the integration tests, i've seen cases where the state was bounded, but the name was not set yet.
so, to make sure that we won't miss this update, I'm checking on both conditions

thotz commented on pkg/operator/ceph/controller/predicate.go:
```diff
@@ -528,6 +516,32 @@ func WatchPredicateForNonCRDObject(owner runtime.Object, scheme *runtime.Scheme)
+				newObj, ok := e.ObjectNew.(*bktv1alpha1.ObjectBucketClaim)
+				if ok {
+					oldObj, _ := e.ObjectOld.(*bktv1alpha1.ObjectBucketClaim)
+					diff := cmp.Diff(
```
`reflect.DeepEqual` can be used to compare the string maps

yuvalif replied:
but i want to compare only changes in labels, not any change in the new/old object

thotz replied:
I mean reflect.DeepEqual on the `label` string map. Not on the entire object

yuvalif replied:
ok. will try that

thotz commented on pkg/operator/ceph/object/notification/controller.go:
```diff
@@ -237,26 +197,42 @@ func (r *ReconcileNotifications) updateCephBucketNotification(notification *ceph
 }
+
+func getCephObjectStoreName(ob bktv1alpha1.ObjectBucket) (types.NamespacedName, error) {
+	// prase the following string: <prefix>-rgw-<store>.<namespace>.svc
```
```suggestion
	// parse the following string: <prefix>-rgw-<store>.<namespace>.svc
```

BlaineEXE commented on pkg/operator/ceph/object/notification/controller_test.go:
```diff
@@ -0,0 +1,502 @@
+			bucketTopic,
+			cephCluster,
+			obc,
+			ob,
```
This is a little confusing because the error appears here, but the fix is above.

BlaineEXE commented on pkg/operator/ceph/object/notification/controller_test.go:
```diff
@@ -0,0 +1,502 @@
+	})
+
+	/*t.Run("provision notification when OB exists", func(t *testing.T) {
+		ob := bktv1alpha1.ObjectBucket{
```
These "pointer receiver" errors can be a little confusing. What this is saying is that an `ObjectBucket` does not match the `DeepCopyObject()` method because it takes a pointer receiver, where the object is not itself a pointer.

`pkg/operator/ceph/object/notification/controller_test.go:460:4: cannot use ob (variable of type "github.com/kube-object-storage/lib-bucket-provisioner/pkg/apis/objectbucket.io/v1alpha1".ObjectBucket) as "k8s.io/apimachinery/pkg/runtime".Object value in array or slice literal: missing method DeepCopyObject (DeepCopyObject has pointer receiver)`

The fix is actually as simple as changing this line to `ob := &bktv1alpha1.ObjectBucket`

thotz commented on pkg/operator/ceph/object/notification/controller.go:
```diff
@@ -237,26 +201,46 @@ func (r *ReconcileNotifications) updateCephBucketNotification(notification *ceph
+			clusterInfo:      clusterInfo,
+			clusterSpec:      &cephCluster.Spec,
 			opManagerContext: r.opManagerContext,
+			owner:            ob.Spec.AdditionalState["cephUser"],
```
```suggestion
			owner:            ob.Spec.AdditionalState[bucket.cephUser],
```

thotz commented on pkg/operator/ceph/object/notification/provisioner.go:
```diff
@@ -79,61 +74,25 @@ func (p *Provisioner) getCephUser(username string, objStore *cephv1.CephObjectSt
-		tlsEnabled,
-	)
-	return session, nil
+	return object.NewS3Agent(accessKey, secretKey, objContext.Endpoint, objContext.ZoneGroup, logger.LevelAt(capnslog.DEBUG), objContext.Context.KubeConfig.CertData)
```
Why it is kept always as debug level for log? 
Will it flood the logs with s3 request details

yuvalif replied:
this is how it is used in: https://github.com/rook/rook/blob/master/pkg/operator/ceph/object/bucket/provisioner.go

leseb commented on pkg/operator/ceph/object/notification/obc_label_controller_test.go:
```diff
@@ -0,0 +1,268 @@
+limitations under the License.
+*/
+
+// Package zonegroup to manage a rook object zone group.
```
```suggestion
// Package notification ...
```

leseb commented on pkg/operator/ceph/object/topic/controller_test.go:
```diff
@@ -0,0 +1,244 @@
+limitations under the License.
+*/
+
+// Package zonegroup to manage a rook object zone group.
```
```suggestion
// Package topic ...
```

leseb commented on pkg/operator/ceph/object/notification/controller_test.go:
```diff
@@ -0,0 +1,427 @@
+limitations under the License.
+*/
+
+// Package zonegroup to manage a rook object zone group.
```
```suggestion
// Package notification ...
```

yuvalif commented:
> This is a CRD so its right place is in WatchControllerPredicate() IMO. Especially since it's called via &source.Kind{Type: &bktv1alpha1.ObjectBucketClaim{}} in the watcher.

@BlaineEXE any comment on that?

BlaineEXE commented:
> > This is a CRD so its right place is in WatchControllerPredicate() IMO. Especially since it's called via &source.Kind{Type: &bktv1alpha1.ObjectBucketClaim{}} in the watcher.
> 
> @BlaineEXE any comment on that?

I think we resolved that Seb's suggestion to keep it in `WatchControllerPredicate()` is more correct, right?

BlaineEXE commented:
Reviewing a 1000 line PR is really challenging. Can we separate the individual changes into smaller PRs that are easier to review and where we can make incremental progress incorporating them into the codebase? I think that will help us prevent ongoing PR fatigue like we had with the initial feature submission.

yuvalif commented:
> Reviewing a 1000 line PR is really challenging. Can we separate the individual changes into smaller PRs that are easier to review and where we can make incremental progress incorporating them into the codebase? I think that will help us prevent ongoing PR fatigue like we had with the initial feature submission.

i don't think that this one is as bad as https://github.com/rook/rook/pull/8426 :-)
but i can break it to separate PRs for refactoring and separate PRs for the unit tests.
the downside is that we would lose the comment history for this one.
let me know what you think.

yuvalif commented:
> > > This is a CRD so its right place is in WatchControllerPredicate() IMO. Especially since it's called via &source.Kind{Type: &bktv1alpha1.ObjectBucketClaim{}} in the watcher.
> > 
> > 
> > @BlaineEXE any comment on that?
> 
> I think we resolved that Seb's suggestion to keep it in `WatchControllerPredicate()` is more correct, right?

removed that commit

leseb commented on pkg/operator/ceph/object/topic/controller_test.go:
```diff
@@ -0,0 +1,244 @@
+			},
+		},
+	}
+	clusterInfo := cephclient.AdminClusterInfo("rook")
```
@yuvalif see below

BlaineEXE reviewed (changes requested):
Looking currently just at commit 1:

I still see a lot of overlap between what the `.../notification/controller.go` reconcile and `.../notification/obc_label_controller.go` reconcile functions are doing. Can we further improve reuse and reduce duplicate (but for some reason slightly differently implemented) code between them?

Of note:
- both check for Topic and ARN (but differently)
- both load cluster info
- both validate object store name

BlaineEXE commented on pkg/operator/ceph/object/topic/controller_test.go:
```diff
@@ -0,0 +1,245 @@
+		expectedARN := "arn:aws:sns:" + store + bucketTopic.Name
+		Create = func(p Provisioner, topic *cephv1.CephBucketTopic) (*string, error) {
+			return &expectedARN, nil
+		}
```
Making `Create` take `Provisioner` as an argument is good use of dependency injection and is helpful for unit testing. :+1:

This doesn't really mock the provisioner but instead mocks the `Create` function. Since there are no unit tests for `Create`, the test itself may very well be missing any bugs that are present in the implementation of `Create`. I would suggest creating thorough unit tests for `Create`, and then this test can check that `Create` is (a) called when it should be and (b) not called when it shouldn't (previous sub-tests). This means that this test doesn't have to (but it can if you like) verify the functionality of `Create`, a sort of unit-integration kind of test.

yuvalif replied:
i plan on testing the provisioner separately. most complexity is in the controller, which is the entity tested here

BlaineEXE commented on pkg/operator/ceph/object/topic/provisioner.go:
```diff
@@ -142,17 +138,12 @@ func createSNSClient(sess *awssession.Session) (snsClient *sns.SNS) {
 }
 
-func (p *Provisioner) Create(topic *cephv1.CephBucketTopic) (*string, error) {
+var Create = func(p Provisioner, topic *cephv1.CephBucketTopic) (*string, error) {
```
Generally, we don't create functions as variables directly. Instead, leave this as a named function, and change usage of `Create` to something like `CreateFunc`, and set `var CreateFunc = Create`.

See example links here:
https://github.com/BlaineEXE/rook/blob/3799542356c7a5651c48f9e48dc9367408963ac4/pkg/operator/ceph/object/controller_test.go#L491-L496
https://github.com/BlaineEXE/rook/blob/0b575703c7313f21d4fd53e3469af9ebdcea0439/pkg/operator/ceph/object/rgw.go#L373-L374

yuvalif replied:
why is this a better mechanism?

yuvalif replied:
done in this commit: https://github.com/rook/rook/pull/9150/commits/5d5905c5f1da46911ed3d94897e10a32c5f7734f

BlaineEXE commented on pkg/operator/ceph/object/topic/provisioner.go:
```diff
@@ -198,7 +193,7 @@ func (p *Provisioner) Create(topic *cephv1.CephBucketTopic) (*string, error) {
 }
 
-func (p *Provisioner) Delete(topic *cephv1.CephBucketTopic) error {
+var Delete = func(p Provisioner, topic *cephv1.CephBucketTopic) error {
```
Ditto

BlaineEXE commented on pkg/operator/ceph/object/topic/provisioner.go:
```diff
@@ -226,7 +221,7 @@ func (p *Provisioner) Delete(topic *cephv1.CephBucketTopic) error {
 }
 
-func GetARN(topic *cephv1.CephBucketTopic) (string, error) {
+var GetARN = func(topic *cephv1.CephBucketTopic) (string, error) {
```
Ditto.

yuvalif commented:
> Looking currently just at commit 1:
> 
> I still see a lot of overlap between what the `.../notification/controller.go` reconcile and `.../notification/obc_label_controller.go` reconcile functions are doing. Can we further improve reuse and reduce duplicate (but for some reason slightly differently implemented) code between them?
> 
> Of note:
> 
> * both check for Topic and ARN (but differently)

fixed - wrapped into one call, and used in both places the same way

> * both load cluster info

fixed - wrapped into one call, and used in both places the same way

> * both validate object store name

this has to be done in different places. in the obc label controller, the object store name is extracted from the OB once,  and compared to the object store name in each topic of each notification. in the notification controller it is extracted from each OB, and validated against the object store name of the topic

yuvalif commented on pkg/operator/ceph/object/notification/provisioner.go:
```diff
@@ -174,10 +124,10 @@ func createS3Events(events []cephv1.BucketNotificationEvent) []*string {
 }
 
-func (p *Provisioner) Create(bucket *bktv1alpha1.ObjectBucket, topicARN string, notification *cephv1.CephBucketNotification, sess *awssession.Session) error {
+func Create(bucket *bktv1alpha1.ObjectBucket, topicARN string, notification *cephv1.CephBucketNotification, s3Agent *object.S3Agent) error {
```
not a hard must for me either, but, i usually prefer syntax like `notification.Create()` over `notification.CreateNotification()`

BlaineEXE replied:
I don't see these functions being called from outside the package, however. Within the package, they are just `Create()` and `DeleteAll()`.  Should these be exported, or are they just internal methods?

yuvalif replied:
done in this commit: https://github.com/rook/rook/pull/9150/commits/5d5905c5f1da46911ed3d94897e10a32c5f7734f

yuvalif commented on pkg/operator/ceph/object/notification/provisioner.go:
```diff
@@ -199,9 +149,9 @@ func (p *Provisioner) Create(bucket *bktv1alpha1.ObjectBucket, topicARN string,
 }
 
-func (p *Provisioner) DeleteAll(bucket *bktv1alpha1.ObjectBucket, sess *awssession.Session) error {
+func DeleteAll(bucket *bktv1alpha1.ObjectBucket, s3Agent *object.S3Agent) error {
```
same as previous comment

yuvalif commented on pkg/operator/ceph/object/notification/controller.go:
```diff
@@ -237,26 +197,46 @@ func (r *ReconcileNotifications) updateCephBucketNotification(notification *ceph
+		return
+	}
+	objectStoreName.Namespace = parsedBucketHost[1]
+	objectStoreName.Name = parsedSubdomain[1]
```
this was my initial implementation, but at some point, the tests started to fail because the integration framework was using a different prefix.
in the latest runs, i don't see that anymore. also seems like this is hardcoded in the `BuildDomainName()` in rgw.go.
so, i will revert the code to assume the hardcoded prefix (and also add a test so we make sure this is won't break i in the future).

leseb commented on pkg/operator/ceph/object/rgw.go:
```diff
@@ -315,6 +317,22 @@ func BuildDomainName(name, namespace string) string {
 }
 
+// ParseDomainName parse the name and namespace from the dns name
+func ParseDomainName(domainName string) (types.NamespacedName, error) {
```
Do we have unit tests for this?

yuvalif replied:
yes. here: https://github.com/rook/rook/blob/5d5905c5f1da46911ed3d94897e10a32c5f7734f/pkg/operator/ceph/object/notification/controller_test.go#L453

BlaineEXE commented on pkg/operator/ceph/object/notification/controller_test.go:
```diff
@@ -54,19 +54,25 @@ var (
 		return nil
 	}
+	return func() {
+		// returned function is doing the cleanup
```
Can we move this function to just above `mockSetup`? This way, it shows up when I hover over the function in VSCode to understand the behavior.

BlaineEXE reviewed (commented):
I have a nit request that should be easy to take care of when squashing all this. Otherwise, looks good to do this and continue revising in follow-up.

thotz reviewed (approved):
LGTM

BlaineEXE commented:
Re-running CI just to make sure TestCephSmokeSuite v1.22 was a flake. Will merge as long as that passes.
